### PR TITLE
CONTRIBUTING: add AI contributions policy

### PR DESCRIPTION
**Pull request by delan**
Following the recent TSC discussion (servo/project#85), this patch proposes a policy for AI contributions.

### CODE REVIEW

**Comment by mrego** [comment]
Just for future reference, this was discussed at https://github.com/servo/project/blob/main/governance/tsc/tsc-2024-04-29.md#ai-generated-prs

File: CONTRIBUTING.md
Comment (gterzian) [inline CONTRIBUTING.md:78]: I would reword this one as a plea to contributors to consider the ethics of their AI use, by:

1. Adding :"We therefore ask contributors to consider these issues before using AI tools to contribute to Servo, especially due to the limit of their current usefulness on a large codebase." 
2. Rephrasing as "It has been reported in the press that certain AI tools require (...), and that their models are built with (...).", and add a few links as examples.

**Review by gterzian — CHANGES_REQUESTED** [review]
I don't think we need to outright ban the use of AI to address the issues with experienced, and I don't think we can enforce such a ban. So I propose to reword this to specifically address problems we experienced(like the maintainer burden created by PRs that appeared to contain generated code which didn't seem to make much sense), and to in other cases point out broader issues to raise awareness of these among contributors.

**Review by SimonSapin — APPROVED** [review]
I agree with the proposed policy as is in the PR at time of review and disagree with suggestions above to weaken it.

I think probabilistic code generation is fundamentally subject to creating subtle bugs, and hand-waving unspecified "reasonable efforts to check correctness" is far from enough.

We may not always be able to "enforce the ban" and detect that a malicious contributor is trying to sneak generated slop through review. That doesn’t mean it’s not worth clearly stating that such behavior is unwelcome.

Authorship concerns are worth mentioning explicitly even if already implied somewhere in the 2426 words of the MPL-2 text.

File: CONTRIBUTING.md
Comment (delan) [reply CONTRIBUTING.md:68]: I would argue that using large language models, by definition, is a failure to make reasonable efforts to check the correctness of our contributions. When we write code, for example, we act as the author. We are in the driver’s seat, deciding how to solve problems and reason about the correctness of the compiler and its output.

This is not the case for the output of generative AI tools, where we act as something more like a reviewer. Being in the passenger seat is always going to put us at a disadvantage, because we have to fight the incentive to accept the output as correct in order to think critically about it.

**Comment by delan** [comment]
I’ve left a reply above, but given the [discussion on Zulip](https://servo.zulipchat.com/#narrow/stream/263399-Technical-Steering-Committee/topic/New.20policy.20proposal.20for.20contributions.20generated.20via.20LLMs), I also want to say that as much as we can debate whether a helpful and ethical generative AI tool could exist someday, our policy should stick to the actual conditions of the present day. If the conditions change, then the policy should change, but until then, we should not allow hypothetical future tools to compromise a policy written to address concrete problems with the concrete tools that exist today.

> We may not always be able to "enforce the ban" and detect that a malicious contributor is trying to sneak generated slop through review. That doesn’t mean it’s not worth clearly stating that such behavior is unwelcome.

I would even say that if anything, the difficulty in reliably enforcing this is a good reason to take a strong and clear stance. Otherwise the wiggle room and uncertainty will make that difficult task impossible.

**Comment by gterzian** [comment]
@delan @SimonSapin Thank for your comments. 

To give you a bit more context, I have used GitHub Co-pilot to write a simple Rust script, using only the standard library, and found it surprisingly useful for boilerplate code(example: generating a `use` statement). I have also tried using Copilot for Servo, as well as doing something with a local model and python, and neither proved very useful in that context(the tool seemed to lack Servo-specific context). I think having an useful AI tool in the context of Servo would require some research, and I was actually hoping the TSC itself could drive some of that. 

I will address the arguments you have put forward, while trying to focus on our contribution policy. 

> We may not always be able to "enforce the ban" and detect that a malicious contributor is trying to sneak generated slop through review. That doesn’t mean it’s not worth clearly stating that such behavior is unwelcome.

Isn't it easy for us to detect problematic PRs, whether they were generated by AI or not? The current discussion originates in the opening of PRs whose content were perceived as being generated with AI, but the main problem was that those PRs were perceived to create an undue maintenance burden by having been poorly prepared. Detecting the problem required no sophisticated tools. On the other hand, it seems impossible to know whether a proper PR included, say, a `use` statement generated by GitHub copilot. Is such a decent PR, created by a contributor who happened to have copilot turned-on in VS code, "unwelcome behavior?" I will answer this rhetorical question below.

> I think probabilistic code generation is fundamentally subject to creating subtle bugs

That is an interesting argument, given the fact that human coding can also be said to be an activity fundamentally subject to creating subtle bugs(and with a much larger empirical data set to back the claim up). What prevents subtle bugs in code is not human skill, but rather automated tools, for example the Rust compiler. On the other hand, preventing subtle bugs in the logic behind the code, whether a formal spec or just an idea in your mind, is a fully human task. So is structuring code in a way that reflects these logical concepts. This leaves plenty of boilerplate coding: similar to code already written, but still too unique to deterministically generate, and so a probabilistic generation seems promising to me.  

> we have to fight the incentive to accept the output as correct in order to think critically about it

This makes sense, but, on the other hand, an equally strong argument could be made that humans will be more critical towards generated code, since it will lack the emotional bound one can feel towards code one wrote by hand. 

> Authorship concerns are worth mentioning explicitly

Then the "Copyright issues" paragraph should be reworded so that it is correct: the project itself cannot be held liable(because of the ["no warranty" part in the license](https://github.com/servo/servo/blob/42061ccafe67ac0440886e48e17a02f8b60525bc/LICENSE#L263)), the risk is on end-users(and perhaps contributors who make false [representations](https://github.com/servo/servo/blob/42061ccafe67ac0440886e48e17a02f8b60525bc/LICENSE#L140)). For a (somewhat dated but still relevant I think)primer: see [this article](https://corporate.findlaw.com/business-operations/the-risks-of-open-source-software.html). The problem is rather the attractiveness of the project for end-users, which could be damaged by a reputation for copyright infringement. 

The [LLVM policy](https://github.com/llvm/llvm-project/pull/91014) on the matter provides a good description of the situation, I think. 

I do think an argument can be made that code generation could result in more inadvertent infringement(although if one goes 20 years back in time, we could replace "code generation" with "using information from search engines"). But, from my experience with Copilot it appears that the snippets that are generated are not specific enough to come directly from a copyrighted source. I also noticed a feature in Copilot that seemed to mitigate this: alerting you if the generated code was a match for existing code on Github(in my case it would then point to forks of Servo, the specific code being a variant of a message enum). I also think it extremely unlikely that there would be a significant chunk of code somewhere on the internet that one could just copy/paste into Servo. I think there is more risk in using a library with an incompatible license(which is not a risk we mitigate in any way, as far as I know). 

> the difficulty in reliably enforcing this is a good reason to take a strong and clear stance

Debatable. I actually think a nudge, such as pointing out the ethical issues, is more likely to result in increased awareness and behavioral changes. 

> Is such a decent PR, created by a contributor who happened to have copilot turned-on in VS code, "unwelcome behavior"? (I will answer this rhetorical question below)

To answer my own question above: I am doubtful that our contribution policy should enforce the choice for contributors of whether to use some kind of AI tool. I thank you for pointing some of these ethical issue out in zulip, I wasn't aware of most of them. But this is a choice that contributors can make for themselves, like the choice of which computer to use, which operating system to install, and which search engine to ask for info from.  

> we should not allow hypothetical future tools to compromise a policy written to address concrete problems with the concrete tools that exist today

We do need to take potential benefits into account, especially given the developing nature of the technology. 
 
I've just read an article entitled ["Mapping the Ethics of Generative AI"](https://arxiv.org/pdf/2402.08323) from the University of Stuttgart, and I'd like to quote a passage(from the "Discussion" paragraph)which I found particularly relevant to our current discussion:

> The literature on the ethics of generative AI is predominantly characterized by a bias towards negative
aspects of the technology, putting much greater or exclusive weight on risks and harms instead
of chances and benefits. This negativity bias is in line with how human psychology is wired (...) can nevertheless result in suboptimal decision-making (...) should be approached with a critical mindset. The numerous benefits and opportunities of adopting generative AI, which may be more challenging to observe or foresee,  are usually overshadowed or discussed in a fragmentary manner in the literature. 

In conclusion: I am not convinced by your arguments, and I think my proposed re-wording would be sufficient to deal with the maintenance problem(which is the only problem the project experienced), and I would like to leave the door open for use of AI within Servo(including research led by the project itself, which may be a prerequisite for developing more useful tools).

**Comment by webbeef** [comment]
While there is no consensus, I think it's pretty clear that only one participant is arguing for IA contributions. And that just looks like moving the Overton window and should be ignored here and not waste the TSC time.

**Comment by delan** [comment]
> While there is no consensus, I think it's pretty clear that only one participant is arguing for IA contributions. And that just looks like moving the Overton window and should be ignored here and not waste the TSC time.

Let’s keep this discussion constructive please.

**Comment by gterzian** [comment]
> One thing that could be relevant for us is policy by mozilla

Here is the Mozilla policy for community contributors in general: https://support.mozilla.org/en-US/kb/contributor-policy-generative-ai-usage

**Comment by sagudev** [comment]
> > One thing that could be relevant for us is policy by mozilla
> 
> Here is the Mozilla policy for community contributors in general: https://support.mozilla.org/en-US/kb/contributor-policy-generative-ai-usage

I was searching for it and couldn't find it (I used the wrong wording). Strangely enough, nobody gave me this link on matrix.

**Comment by gterzian** [comment]
> allowing people to “contribute” large volumes of nonsense that statistically resembles code.
> at the expense of the wellbeing of our reviewers and the correctness and legality of our work.

Off course I cannot argue in favor of either of these harms, but I see what mean.  

> These are well-documented phenomena

Do you have a link perhaps? 

----------------------

> We should discuss this at TSC level and see if we can reach some kind of agreement somehow.

Perhaps as a last try here, see below.

> Reviews are already time-consuming, and it’s rare for us to reject a PR outright. 

Allright, so then if we do stick with the overall ban, how about we add to "we may decide to revise this policy at a later date" something like "and we are open to proposals of beneficial use cases from our community", just to be clear that the TSC will remain open-minded about potential beneficial uses of AI, with a commitment to spend some time reviewing potential tools if brought forward by either members of our community or TSC members? That is the minimal change that I think changes the tone of the policy and would make it easier to people to actually present such beneficial use cases going forward.

**Comment by fabricedesre** [comment]
> > > One thing that could be relevant for us is policy by mozilla
> > 
> > 
> > Here is the Mozilla policy for community contributors in general: https://support.mozilla.org/en-US/kb/contributor-policy-generative-ai-usage
> 
> I was searching for it and couldn't find it (I used the wrong wording). Strangely enough, nobody gave me this link on matrix.

Note that this is for community contributors to the *support* knowledge base of Mozilla. I don't think it applies to code contributions - there would certainly be details about potential IP issues if that was applicable to code.

**Comment by fabricedesre** [comment]
> Also please change "these contributions could make the project liable for copyright infringement" into something along the lines of "these contributions could lessen the appeal of the project to end-user from a copyright protection perspective".

Who are the "end users" here? Are they developers re-using Servo, or people using a Servo based product?

**Comment by gterzian** [comment]
> Who are the "end users" here? 

The "You" in https://github.com/servo/servo/blob/c24183096f8714ea28d592f34db6997d45a22a33/LICENSE#L271

**Comment by delan** [comment]
> Allright, so then if we do stick with the overall ban, how about we add to "we may decide to revise this policy at a later date" something like "and we are open to proposals of beneficial use cases from our community", just to be clear that the TSC will remain open-minded about potential beneficial uses of AI, with a commitment to spend some time reviewing potential tools if brought forward by either members of our community or TSC members? That is the minimal change that I think changes the tone of the policy and would make it easier to people to actually present such beneficial use cases going forward.

8192d28b0994c39737346cce9b399bf60c1c00f6 clarifies that we are open to revising the policy, and makes it more explicit that proposals can be made for particular tools that mitigate the effects outlined in the policy.

> Also please change "these contributions could make the project liable for copyright infringement" into something along the lines of "these contributions could lessen the appeal of the project to end-user from a copyright protection perspective".

I’m not so sure about this change. It seems to me that the new wording is less clear than the old one, and I don’t understand how “copyright protection” appeals to the end user, but maybe I’ve misunderstood something.

**Comment by mrobinson** [comment]
Do we have to answer this question definitively? Why not say something like "these contributions might contain copyright violations?"

**Comment by mrobinson** [comment]
Do we have to answer this question definitively? Why not say something like "these contributions might contain copyright violations?"

**Comment by gterzian** [comment]
> Or have I misunderstood what you were referring to by “explanation above”?

You have understood it correctly, in that I am referring to the Disclaimer of Warranty. 

> I believe we can still be held liable for copyright infringement

What is this belief based on? 

I am not a lawyer either, and this is not legal advice; I did do a bit of research before writing my comments above and suggest the revised wording. For example, here is an article that goes through the warranty clause of a similar license: https://writing.kemitchell.com/2016/09/21/MIT-License-Line-by-Line

Also, the wording of the license seems clear to me:

> Covered Software is provided under this License on an "as is" basis, without warranty of any kind (...) including, (...) warranties that the Covered Software is (...) non-infringing. The entire risk as to the quality and performance of the Covered Software is with You.

------

> Do we have to answer this question definitively? Why not say something like "these contributions might contain copyright violations?"

We don't have to answer this question definitely(how could we?), but if concerns about copyright infringement are one of the reasons for this policy, then in the spirit of informed policy-making we should understand how copyright infringement can affect the project and reflect this in the wording of the policy.

**Comment by delan** [comment]
681d999afb95a1d59a4111d3bd76f6344865aa1b rewords the point in a way that should sidestep this question.

**Comment by gterzian** [comment]
> [681d999](https://github.com/servo/servo/commit/681d999afb95a1d59a4111d3bd76f6344865aa1b) rewords the point in a way that should sidestep this question. Since neither of us are qualified to answer this question definitively, this is probably not the best use of our time.

I think it is a good use of our(at least my) time, since our role here is to manage an open-source project and a good understanding of its license seems useful for that purpose. We don't have to be qualified to answer this question definitively because we are not giving anyone legal advice here, it is only own understanding that could be clarified. That is my personal opinion, and anyone who shares it can read the above arguments and the document in the links. 

I will re-read the whole thing on Monday, and approve it before the TSC; not because my approval is necessary--we could [vote](https://github.com/servo/project/blob/main/governance/CHARTER.md#3--tsc-voting) and I am sure it would be approved with or without my vote--but rather because that is how PRs in Github seem to work(a request for changes needs an approval).

**Review by gterzian — APPROVED** [review]

**Pull request merged**
